src/Amalgam/BinaryPacking.h

```diff
@@ -1,24 +1,109 @@
 #pragma once
 
-//project headers:
-#include "HashMaps.h"
-
 //system headers:
+#include <limits>
 #include <string>
+#include <utility>
 #include <vector>
 
-typedef uint64_t OffsetIndex;
-
 typedef std::vector<uint8_t> BinaryData;
 
-//Appends the offset index oi to BinaryData
-void UnparseIndexToCompactIndexAndAppend(BinaryData &bd_out, OffsetIndex oi);
+//Huffman Encoding implementation for compressing and decompressing data
+template <typename value_type>
+class HuffmanTree
+{
+public:
+	constexpr HuffmanTree(value_type value, size_t value_frequency, size_t node_index,
+		HuffmanTree<value_type> *left = nullptr, HuffmanTree<value_type> *right = nullptr)
```

Comment:
Does using `std::unique_ptr<T>` make sense here?  Of note, that has the property that, if the unique pointer goes out of scope holding a value, it automatically deletes the value it owns (and you could remove the destructor).